tensorboard/webapp/app_routing/effects/app_routing_effects.ts

```diff
@@ -102,6 +102,21 @@ type NavigationOptions = {
   namespaceUpdate: NamespaceUpdate;
 };
 
+const generate32bitRandomId = () => {
+  let arr = new Uint8Array(32);
+
+  // The usage of Crypto `getRandomValues` does not require return an array.
+  // This is for the convenience of mock the funciton in testing.
+  arr = crypto.getRandomValues(arr);
```

Comment:
Did you consider using callFake instead?

For example:

        spyOn(window.crypto, 'getRandomValues').and.callFake((arr) => {
          if (!(arr instanceof Uint8Array)) {
            return arr;
          }
          arr[0] = TEST_UINT8ARRAY[0];
          arr[1] = TEST_UINT8ARRAY[1];
          arr[2] = TEST_UINT8ARRAY[2];
          arr[3] = TEST_UINT8ARRAY[3];
          return arr;
        });

It would remove the awkwardness here at the cost of making your tests more verbose (although you could probably minimize the verbosity with a better callFake() implementation)

Reply:
callFake was my first attempt and makes more sense to me. Unfortunately I was stock at weird typescript error.
`getRandomValues` is a generic function type `<T extends ArrayBufferView | null >(arr: T) => T`
see the first commit of this pr. both `set`, `fill`, `index` are not compilable.

Reply:
These lines should fix that for you by telling the compiler to narrow the type:

```
      if (!(arr instanceof Uint8Array)) {
        return arr;
      }
```